Add a scheduler running the system functions once at each frame.

catageek opened the pull request:
Each system registered has a dedicated thread that executes its 2 main functions every 1/60 second.

When not running the system functions, the scheduler assigns pending tasks to threads. The tasks can be split in small blocks to allow collaborative behaviour (the NetworkController runs entirely asynchronusly using the scheduler). The keyboard and mouse controllers should us asynchronous tasks only.

For illustration, a FakeSystem has been added, and its 2 functions implemented to print a message each time they are executed.

The scheduler uses a constant number of threads and limits the number active threads. The implementation is still a work in progress, but the interface should not change.

Commit:
Change folder hierarchy

Commit:
Update include paths

Commit:
Add scheduler

Commit:
Add FakeSystem for demo

Commit:
Enable multithreading in gcc

Commit:
Fix argument of functor

Commit:
Revert commit d40b0592e9abc3a3122ea145237ca26178b57166

Commit:
Add multithread flag to gcc

Commit:
Cut lines to 80/120 cols

adam4813 commented on include/System.h:
```diff
@@ -0,0 +1,43 @@
+         * prepared and stored temporarily to be retrieved by RunBatch().
+         *
+         */
+        virtual void HandleEvents(const std::chrono::time_point<std::chrono::steady_clock, frame_unit>& timepoint) = 0;
```
So, this would be equivalent to the Update method from Sigma. in a general sense.

catageek replied:
HandleEvents() followed by RunBatch() is equivalent to Update(), but split in 2 parts.

adam4813 commented:
I'll give this a windows test tomorrow hopefully and we can merge it in assuming it'll work with the design.

adam4813 commented:
Ok...so for some reason on Visual Studio steady_clock::now() calls system_clock::now() which can't be compared to a steady_clock such as in the line 32.

Also Visual Studio doesn't have thread_local, but it does have __declspec(thread). Unfortunately __declspec(thread) doesn't work for non-POD.

Commit:
Remove thread_local storage

catageek commented:
I removed thread_local and use local variables instead.

For the clock issue, I need to implement something different, but it does not prevent the merge since it is an internal issue.

EDIT : We will use the GLFW clock function once GLFW and this PR will be merged

Commit:
Merge remote-tracking branch 'origin/develop' into feature-scheduler

Conflicts:
	src/main.cpp

Commit:
Add GLFW OS instance to TrillekGame

Commit:
Use GLFW time function as clock and implement a global terminate flag

adam4813 commented on include/TrillekScheduler.h:
```diff
@@ -0,0 +1,218 @@
+        virtual void RunTask() = 0;
+
+        bool IsNow() const {
+            return timestamp < steady_clock::now();
```
This line fails to compile on visual studio because steady_clock::now() calls system_clock::now() which for some reason can't be compared even though steady_clock is a subclass of system_clock.

adam4813 commented on include/TrillekScheduler.h:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        void Reschedule(frame_unit&& delay) {
+            timestamp = steady_clock::now() + delay;
```
Same issue as above, but with attempting to add a system_clock to a steady_clock.

Commit:
Replace steady_clock::now() calls by GLFW timer in header

catageek commented:
I forgot to replace the calls of now() in header. It is fixed.

Commit:
OS loop is managed by the scheduler.

The main thread is blocked in a loop, waiting the window to be closed.
The terminate flag has been replaced by a call to OS::Closing().

Commit:
Manage termination of the program

Commit:
Systems have a Terminate() function called when the program exits

catageek commented:
We have now a clean program termination.

Commit:
Add some comments

catageek commented:
Any reference to steady_clock will be replaced by system_clock in time_point definition, since the only value that time_point wants to access is the epoch of the clock. I hope that MSVC will be happy.

Also, I will try to cut lines to be more readable...

Commit:
Remove all steady_clock reference as workaround of MSVC bug

milesrout commented:
@adam4813 Merge this.

adam4813 commented:
Why, when it breaks all visual c++ compiling?

I am still not certain as to what multithreading at this stage gets us. Unless we are utilizing out of order logic processing for each system on separate cores, threading won't grant us any major benefits. 

Tomorrow I will try and sit down with @catageek and determine a way to get this to compile via visual c++ as we will be excluding a chunk of potential modders that are windows lovers such as myself.

milesrout commented:
If it breaks then fix it, @catageek can't compile on Windows as you have known for weeks.

milesrout commented:
If you don't build around being thread-safe from the start it will be very difficult to add it in later. You know that, I don't know why I'm having to tell you this.

Commit:
Remove tests that are specific to the c++ library used

Commit:
Restrict atomic_queue definition to local scope

Commit:
Fix indentation

Commit:
Remove GLFW_DEBUG_LIBRARY variable

Commit:
Merge branch 'develop' of https://github.com/trillek-team/trillek-client-core into develop

Commit:
Merge branch 'develop' into feature-scheduler

Conflicts:
	include/OS.h
	src/OS.cpp
	src/main.cpp

Commit:
Attach the window context to the graphic system thread

Commit:
Fix errors when building with VS2013

Commit:
Find workaround on VS2013 bug

Commit:
Rename files and fix indentation

catageek commented:
Ready to be merged to the current develop branch and VS2013-compatible. The window control is now transferred to the graphic system thread (OpenGL is not multithread-friendly) and returned to the main thread when closing.

Theoretically, all the graphic stuff, including the window creation, should be moved to the graphic system to have a trivial main().

I have created an ThreadInit function in the SystemBase interface that is executed once the system has its own thread. It is used to make the context of the window current to the thread. All the Initialize() function may be moved to it.

Basically, nothing should be initialized in main except the scheduler with the list of systems to start, and the controllers tasks already queued.

I left things in this state because I am not sure of the status of the "transform" and "resource" instances created in main, and if the can/must be moved.

milesrout commented:
@catageek VS2013 with the free CTP supports rvalue references, you don't have to remove them. 

and.. uhh.. did you just make a whole lot of changes in that last commit that had already been made elsewhere in the repo? That's what a merge is for. 

Also the endless `typedef`s make the code very difficult to read.

catageek commented:
@milesrout I spent hours to make this thing built by VS without errors, I leave the improvements to people who have VS and don't have to wait 5 minutes to get the result of the build like in appveyor.

I don't think I retyped some changes made in other commits. I merged the current develop branch to work on the latter version.

And typedefs are here to respect the holy limit of 80 chars on each line.

milesrout commented:
If the code requires typedefs to respect 80 chars per line then it's too
complicated and should be refactored into smaller functions. If you can
read it on one screen then it's too complicated and should be refactored
into smaller functions. That's part of the benefit of auto, in fact.

Nice work, though, nice work.

On Fri, Jun 6, 2014 at 8:28 PM, catageek notifications@github.com wrote:

> @milesrout https://github.com/milesrout I spent hours to make this
> thing built by VS without errors, I leave the improvements to people who
> have VS and don't have to wait 5 minutes to get the result of the build
> like in appveyor.
> 
> I don't think I retyped some changes made in other commits. I merged the
> current develop branch to work on the latter version.
> 
> And typedefs are here to respect the holy limit of 80 chars on each line.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/trillek-team/trillek-client-core/pull/13#issuecomment-45314052
> .

Commit:
Call OSmessageLoop() from the main thread

Commit:
Fix race condition in scheduler main loop
